File: Day10-19/18.py
from collections import defaultdict


class VM:
    def __init__(self, instructions, registers=None, part1=False):
        self.instructions = instructions
        self.pc = 0
        self.send_val = None
        self.receive_queue = []

        if registers is None:
            registers = defaultdict(int)
        self.registers = registers

        self.part1 = part1
        self.first_receive = None

    def _get_value(self, operand):
        try:
            return int(operand)
        except:
            return self.registers[operand]

    def run_instruction(self):
        inst = instructions[self.pc]
        if inst[0] == 'snd':
            self.last_sound = self._get_value(inst[1])
            self.send_val = self._get_value(inst[1])
        elif inst[0] == 'set':
            self.registers[inst[1]] = self._get_value(inst[2])
        elif inst[0] == 'add':
            self.registers[inst[1]] += self._get_value(inst[2])
        elif inst[0] == 'mul':
            self.registers[inst[1]] *= self._get_value(inst[2])
        elif inst[0] == 'mod':
            self.registers[inst[1]] %= self._get_value(inst[2])
        elif inst[0] == 'rcv':
            if self.part1:
                self.first_receive = self.last_sound
            else:
                if self.receive_queue:
                    self.registers[inst[1]] = self.receive_queue.pop()
                else:
                    return
        elif inst[0] == 'jgz':
            if self._get_value(inst[1]) > 0:
                self.pc += self._get_value(inst[2])
                return
        self.pc += 1


def make_process_reg(pid):
    reg = defaultdict(int)
    reg['p'] = pid
    return reg


def run_vm_pair(instructions):
    P0 = VM(instructions, make_process_reg(0))
    P1 = VM(instructions, make_process_reg(1))
    part2 = 0
    pc0, pc1 = -1, -1
    while P0.pc != pc0 or P1.pc != pc1:
        pc0 = P0.pc
        pc1 = P1.pc
        if P0.send_val is not None:
            P1.receive_queue = [P0.send_val] + P1.receive_queue
            P0.send_val = None
        if P1.send_val is not None:
            P0.receive_queue = [P1.send_val] + P0.receive_queue
            P1.send_val = None
            part2 += 1
        P0.run_instruction()
        P1.run_instruction()
    return part2

if __name__ == '__main__':
    with open('18.txt') as f:
        instructions = [l.split() for l in f.readlines()]
    P1 = VM(instructions, part1=True)
    while P1.first_receive is None or P1.first_receive == 0:
        P1.run_instruction()
    print("Part 1: {}".format(P1.first_receive))

    part2 = run_vm_pair(instructions)
    print("Part 2: {}".format(part2))

File: Day10-19/18.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type VM struct {
	instructions *[][]string
	pc           int64
	registers    map[string]int64

	sendFlag bool
	sendVal  int64

	recvQueue []int64

	part1             bool
	firstReceivedFlag bool
	firstReceived     int64
	lastPlayed        int64
}

func (vm *VM) getValue(token string) int64 {
	val, err := strconv.Atoi(token)

	if err == nil {
		return int64(val)
	}

	if _, ok := vm.registers[token]; !ok {
		vm.registers[token] = 0
	}
	return vm.registers[token]
}

func (vm *VM) runInstruction() {
	inst := (*vm.instructions)[vm.pc]

	switch inst[0] {
	case "snd":
		vm.lastPlayed = vm.getValue(inst[1])
		vm.sendVal = vm.getValue(inst[1])
		vm.sendFlag = true
	case "set":
		vm.registers[inst[1]] = vm.getValue(inst[2])
	case "add":
		prev := vm.getValue(inst[1])
		vm.registers[inst[1]] = prev + vm.getValue(inst[2])
	case "mul":
		prev := vm.getValue(inst[1])
		vm.registers[inst[1]] = prev * vm.getValue(inst[2])
	case "mod":
		prev := vm.getValue(inst[1])
		if vm.getValue(inst[2]) == 0 {
			break
		}
		if prev < 0 {
			prev = -prev
		}
		vm.registers[inst[1]] = prev % vm.getValue(inst[2])
	case "rcv":
		if vm.part1 {
			if vm.getValue(inst[1]) != 0 {
				vm.firstReceived = vm.lastPlayed
				vm.firstReceivedFlag = true
			}
		} else if len(vm.recvQueue) > 0 {
			vm.registers[inst[1]] = vm.recvQueue[0]
			vm.recvQueue = vm.recvQueue[1:]
		} else {
			return
		}
	case "jgz":
		if vm.getValue(inst[1]) > 0 {
			vm.pc += vm.getValue(inst[2])
			return
		}
	default:
		panic("Invalid operation")
	}
	vm.pc++
}

func NewVM(instructions *[][]string, registers map[string]int64, part1 bool) VM {
	return VM{
		instructions: instructions,
		pc:           0,
		registers:    registers,

		sendFlag: false,
		sendVal:  0,

		recvQueue: make([]int64, 0),

		part1:             part1,
		firstReceivedFlag: false,
		firstReceived:     0,
		lastPlayed:        0,
	}
}

func makeProcessRegister(pid int64) map[string]int64 {
	reg := make(map[string]int64)
	reg["p"] = pid
	return reg
}

func runVMPair(instructions *[][]string) int {
	P0 := NewVM(instructions, makeProcessRegister(0), false)
	P1 := NewVM(instructions, makeProcessRegister(1), false)

	part2 := 0
	pc0 := int64(-1)
	pc1 := int64(-1)
	for P0.pc != pc0 || P1.pc != pc1 {
		pc0 = P0.pc
		pc1 = P1.pc

		if P0.sendFlag {
			P1.recvQueue = append(P1.recvQueue, P0.sendVal)
			P0.sendFlag = false
		}
		if P1.sendFlag {
			P0.recvQueue = append(P0.recvQueue, P1.sendVal)
			P1.sendFlag = false
			part2++
		}

		P0.runInstruction()
		P1.runInstruction()
	}
	return part2
}

func main() {
	dat, _ := ioutil.ReadFile("./18.txt")
	lines := strings.Split(string(dat), "\n")

	instructions := make([][]string, len(lines))
	for i := 0; i < len(lines); i++ {
		instructions[i] = strings.Fields(lines[i])
	}

	P1 := NewVM(&instructions, make(map[string]int64), true)
	for !P1.firstReceivedFlag || P1.firstReceived == 0 {
		P1.runInstruction()
	}
	fmt.Printf("Part 1: %d\n", P1.firstReceived)

	part2 := runVMPair(&instructions)
	fmt.Printf("Part 2: %d\n", part2)
}
